packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/api/SelectionDialogDescriptionConverter.java

```diff
@@ -0,0 +1,341 @@
+                    .map(parameters -> parameters.get(TREE_DESCRIPTION_ID).get(0))
+                    .filter(treeDescriptionIdParameter -> treeDescriptionId.equals(treeDescriptionIdParameter))
+                    .isPresent();
+        };
+        String treeId = SELECTION_PREFIX + "?" + TREE_DESCRIPTION_ID + "=" + treeDescriptionId;
+        return TreeDescription.newTreeDescription(treeDescriptionId)
+                .label(DIALOG_DESCRIPTION_TREE_REPRESENTATION_NAME)
+                .idProvider(variableManager -> variableManager.get(GetOrCreateRandomIdProvider.PREVIOUS_REPRESENTATION_ID, String.class).orElse(treeId))
+                .canCreatePredicate(canCreatePredicate)
+                .childrenProvider(childrenProvider)
+                .deletableProvider(variableManager -> false)
+                .deleteHandler(deleteHandler)
+                .renameHandler(renameHandler)
+                .editableProvider(variableManager -> false)
+                .elementsProvider(elementsProvider)
+                .hasChildrenProvider(hasChildrenProvider)
+                .treeItemIdProvider(treeItemIdProvider)
+                .kindProvider(kindProvider)
+                .labelProvider(labelProvider)
+                .targetObjectIdProvider(targetObjectIdProvider)
+                .iconURLProvider(imageURLProvider)
+                .selectableProvider(isSelectableProvider)
+                .treeItemObjectProvider(this::getTreeItemObject)
+                .parentObjectProvider(this::getParentObject)
+                .build();
+    }
+
+    private Boolean hasChildren(AQLInterpreter interpreter, SelectionDialogTreeDescription selectionDialogTreeDescription, VariableManager variableManager,  Function<VariableManager, Boolean> isSelectableProvider) {
+        boolean hasChildren = false;
+        Object self = variableManager.getVariables().get(VariableManager.SELF);
+        hasChildren = !this.computeChildrenFromExpression(variableManager, interpreter, selectionDialogTreeDescription).isEmpty();
+        return hasChildren && this.hasCompatibleDescendants(interpreter, selectionDialogTreeDescription, variableManager, self, false, isSelectableProvider);
+    }
+
+    private boolean hasCompatibleDescendants(AQLInterpreter interpreter, SelectionDialogTreeDescription selectionDialogTreeDescription, VariableManager variableManager, Object object, boolean isDescendant, Function<VariableManager, Boolean> isSelectableProvider) {
+        VariableManager childVariableManager = variableManager.createChild();
+        childVariableManager.put(VariableManager.SELF, object);
+        return isDescendant && isSelectableProvider.apply(childVariableManager)
+                || this.computeChildrenFromExpression(childVariableManager, interpreter, selectionDialogTreeDescription).stream().anyMatch(eContent -> this.hasCompatibleDescendants(interpreter, selectionDialogTreeDescription, childVariableManager, eContent, true, isSelectableProvider));
+    }
```

Comment:
So for you it is the specifier responsibility to define this rules?
It can be complicated to implement in the studio

Reply:
Let's consider the following tree:

```
- A
  - B
    - C
      - D
        - E
          - F (with 1 million elements under F)
```

For every single semantic change in the workbench, for each tree item displayed (here A, B, C, D, E, F) we will execute `hasChildren` (so 6 times per refresh) which will each perform an `eAllContents` on the whole tree just because we want to be smarter than the specifier who has told us to display something by trying to not display potentially useless stuff. That's a massive performance penalty for everybody on the whole project just because ONE person has opened the selection dialog.

If the specifier tells us to display something, we display it. I can understand the intention behind this feature but we can't accept such a performance it for something like that. On top of it, I disagree with the need for such feature. One may need to see the children of an element in order to decide which was should be selected. I don't see why we should decide to hide something that the specifier has explicitly told us to display thanks to the `childrenExpression`. If the specifiers wants to perform `eAllContents` everywhere that's their choice not ours.

Reply:
I understand the arguments. For information, I used the same logic we already have in org.eclipse.sirius.components.collaborative.widget.reference.browser.ModelBrowsersDescriptionProvider.

Reply:
(●__● )

In that case, I may have missed something there. I'll look into it...

( ｀▽´)Ψ

Reply:
FWIW this behavior in ModelBrowsersDescriptionProvider was added explicitly via #2171 (commit https://github.com/eclipse-sirius/sirius-web/commit/a76a04be9e342f5c97e8cf7ffcda6d93c93e14c3). I wrote the issue (and PR), but form what I remember this was originally from user feedback on the first version.